### DISCUSSION TITLE
Encoding cue message to specified text protobuf message

### DISCUSSION THREAD

**Discussion by jdb** [cue-lang/cue#1193] [Q&A]
Hello, I saw git commits about decoding and encoding text protobuf messages, I think the commits means that item 2 and item 3 of the section "Support is planned for" in  https://cuelang.org/docs/integrations/protobuf/
are implemented. Did I get this right?

I read the [encoding/protobuf/textproto/encoder.go][encoder]  module and I did not find the method I was expecting: I thought there would be a method that encode to text protobuf, given the following args:
- a cue value to serialize,
- **the name of a proto message to serialize to**, 
- **a proto schema definition for the message**.

Instead, I think the input to the Encode method is a Cue value, which only gets reformatted in the text protobuf syntax and into bytes but without validating against a message schema.

Q1. Is there an example I can look at that serialize a cue value to a specified proto message and reporting an error if the cue value have extra fields missing from the schema, or field values are of a different type than the one in the schema?

I guess I can build that validation myself right after the value got encoded and try to decode it with the standard textproto parser: 
Q2. is this the idea or other scenarios are envisioned?

Thanks for your patience, I am a noob now, I just read the docs and installed cue.

[encoder]: https://github.com/cuelang/cue/blob/2362858826cd0a63fd33691928665c60c3411361/encoding/protobuf/textproto/encoder.go#L51
[option]: https://github.com/cuelang/cue/blob/2362858826cd0a63fd33691928665c60c3411361/encoding/protobuf/textproto/encoder.go#L38

**Comment by myitcv** [comment]
Hi @jdb - welcome to the CUE project! Apologies for the delay in replying.

> Did I get this right?

Yes, that matches my understanding. But I will cc @mpvl for a definitive answer.

> Instead, I think the input to the Encode method is a Cue value, which only gets reformatted in the text protobuf syntax and into bytes but without validating against a message schema.

Can you share an example of what you are trying to do, with inputs etc? That will help providing a concrete answer on how to use the existing API, or whether there is indeed a gap. 

In particular I'm not clear what you mean by a "CUE message" - are you referring to a definition? 

> Thanks for your patience, I am a noob now, I just read the docs and installed cue.

Absolutely no problem at all! It's great to work through these sorts of things, because it helps us to understand where the API/docs etc can be improved, so thank you for raising the discussion!

**Comment by jdb** [comment]
Thanks for you answer Paul, I can follow up when I have something setup. 

The common problem I am to solve designing config files that share a lot of similarities. I see many config language using inheritance, and this is not working perfectly for us. I want to experiment with Cue where the merge is associative, commutative and idempotent. Hopefully, I can write  many Cue config files that join with a common config file to denormalize each configs into fully expanded, redundant structs. 

Protobuf is ubiquitous where I work, so likely I will need to :
1. merge in data in the protobuf format, 
2. or convert the end format to protobuf message so they can be used by the systems they configure.

I took the liberty to edit my previous to substitute "Cue message" by "Cue value", I did not want to imply a Cue definition.

**Reply by myitcv** [reply]
This all sounds very doable and firmly in the wheelhouse of CUE. If you have a concrete example, then happy to help work through this.
